fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/metadata/JoinedRecordTypeBuilder.java

```diff
@@ -167,22 +213,59 @@ public Join addJoin(@Nonnull String left, @Nonnull KeyExpression leftExpression,
             throw new RecordCoreArgumentException("Two sides of join are not the same size and will never match")
                     .addLogInfo("left", leftExpression, "right", rightExpression);
         }
-        Join join = new Join(left, leftExpression, right, rightExpression);
+        Join join = new Join(left, leftExpression, leftExpression, right, rightExpression, rightExpression);
+        joins.add(join);
+        return join;
+    }
+
+    /**
+     * Add a new equi-join between the specified left and right sides.
+     *
+     * @param left the name of the type on the LHS of the join.
+     * @param leftIndexExpression the KeyExpression to use when constructing indexes. The entries in this KeyExpression
+     * must line up correctly with how the join index will store records on disk.
+     * @param leftJoinExpression the KeyExpression to use when applying the join condition on incoming records from the
+     * left-hand-side. This
+     * expression isn't stored, it's just used for join comparisons (For example, when applying a join condition where
+     * the join is between two projections).
+     * @param right the name of the type on the right hand side of the join.
+     * @param rightIndexExpression The KeyExpression to use when constructing indexes. The entries in this KeyExpression
+     * must line up correctly with how the join index will store RHS records on disk.
+     * @param rightJoinExpression the KeyExpression to use when applying the join conditions on incoming records from
+     * the right-hand-side.
+     * This expression isn't stored, it's just used for join operations (For example, when applying a join condition
+     * where
+     * the join is between two projections).
+     * @return a representation of the join.
+     */
+    @Nonnull
+    public Join addJoin(@Nonnull String left, @Nonnull KeyExpression leftIndexExpression, @Nonnull KeyExpression leftJoinExpression,
+                        @Nonnull String right, @Nonnull KeyExpression rightIndexExpression, @Nonnull KeyExpression rightJoinExpression) {
+        if (leftIndexExpression.getColumnSize() != rightIndexExpression.getColumnSize()) {
+            throw new RecordCoreArgumentException("Two sides of join are not the same size and will never match")
+                    .addLogInfo("left", leftIndexExpression, "right", rightIndexExpression);
+        }
+        //TODO(bfines) validate column sizes for the value expressions also
```

Comment:
If I'm understanding this PR correctly, are the actual requirements that `leftIndexExpression.getColumnSize() == rightJoinExpression.getColumnSize()` and `leftJoinExpression.getColumnSize() == rightIndexExpression.getColumnSize()`? Because I _think_ the only things that get compared are the index expressions from one side and the join expressions from the other.

Reply:
that is correct--I adjusted the column size checks to account for this